app/src/main/java/com/arctouch/codechallenge/AppApplication.kt

```diff
@@ -0,0 +1,28 @@
+package com.arctouch.codechallenge
+
+import android.app.Application
+import com.arctouch.codechallenge.api.TmdbApi
+import com.arctouch.codechallenge.util.Constants.Companion.URL
+import okhttp3.OkHttpClient
+import okhttp3.logging.HttpLoggingInterceptor
+import retrofit2.Retrofit
+import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
+import retrofit2.converter.moshi.MoshiConverterFactory
+
+class AppApplication : Application() {
```

Comment:
A better practice is to encapsulate all network features (logging, api setup, etc) outside any Android contexts, like `Application`, `Activity` or other application-level component. A better alternative for this can be accomplished by moving the network code to another class, like `BaseApi` and make `TmdbApi` inherit all its behaviour needed. In that way, `logging` should be instantiated lazily using the lazy delegate in Kotlin. An example of how your code should be rewritten is below:

    class BaseApi {
        protected val logging by lazy { HttpLoggingInterceptor() }
        protected val okHttpClient by lazy { OkHttpClient.Builder().addInterceptor(logging).build() }
        protected api by lazy { Retrofit.Builder()
                      .baseUrl(URL)
                     .client(okHttpClient)
                     .addConverterFactory(MoshiConverterFactory.create())
                     .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                     .build() }     
         init {
             logging.level = HttpLoggingInterceptor.Level.BODY
         }
     }
 }

Doing this, you will have those variables instantiated only when they are first referenced.

So, you can than make `TmdbApi` to be accessed by another data source abstraction, like `TmdbRepository`:

    class TmdbRepository(private val tmdbApi: TmdbApi) {
        fun loadMovies() = tmdbApi.loadMovies()
        // other codes go here
    }

And make the ViewModel to access the Repository. And the `AppApplication` does not need to know about anything related to networking.

For further reading, I suggest you to read this article that explains more about how to structure better an architecture using Architecture Components: https://medium.com/mindorks/android-architecture-components-room-viewmodel-and-livedata-50611793e4a9

Reply:
yes, you're right I could have done it that way and could still do better. Instead of creating a TmdbRepository class I could leave it as an interface and to create a class called TmdbStore where I would implement the loadMovies() method, so this way I will keep a contract for the call to the server.